Feature/udp stream rx

bmagistro opened the pull request.

bmagistro committed:
l4p/udp: add establish and stream rx support

Signed-off-by: Ben Magistro <koncept1@gmail.com>
Change-Id: Ibaa3fe0dfabc591908883a0091e5de0723b8a1fa

bmagistro committed:
l4p/udp: add udp stream rx size

This adds the ability to query the udp rx queue size. There is no
equivalent avaiable for tx in udp.

Signed-off-by: Ben Magistro <koncept1@gmail.com>
Change-Id: Ic68e870d6db0297c65cc17065bffb8828a51a0cc

ttosbt commented on lib/l4p/udp_rxtx.c:
```diff
@@ -108,6 +108,33 @@ rx_stream(struct tle_udp_stream *s, void *mb[], struct rte_mbuf *rp[],
+	// rwl_release(&s->rx.use);
+
+	return n;
+
```
Remove this extra white space.

ttosbt commented on lib/l4p/udp_stream.c:
```diff
@@ -276,6 +276,74 @@ tle_udp_stream_open(struct tle_ctx *ctx,
 	return &s->s;
 }
 
+	struct tle_stream *
```
The breaking of this definition seems wonky...

bmagistro replied:
Their style...not mine.

bmagistro replied:
the leading space will be removed

ttosbt commented on lib/l4p/udp_stream.c:
```diff
@@ -276,6 +276,74 @@ tle_udp_stream_open(struct tle_ctx *ctx,
+	/* setup L4 port info */
+	s->s.port.src = ((const struct sockaddr_in *)&prm->remote_addr)->sin_port;
+	s->s.pmsk.src = UINT16_MAX;
+	s->s.port.dst = ((const struct sockaddr_in *)&prm->local_addr)->sin_port;
```
Nit: Above you've used the ordering dst,src but here you've switched it up.

bmagistro replied:
most of it was ordered as it was pulled from other parts of code to make it work.  should there be some standard ordering...probably

ttosbt replied:
Gotcha...you know me, haha...that OCD kicks in.

bmagistro replied:
well this is fixed but for other reasons.

ttosbt commented on lib/l4p/tle_udp.h:
```diff
@@ -98,6 +98,19 @@ int
+ *   Stream to retrieve state information from.
+ * @return
+ *   On success, the value will be >= 0 and can safely be stored
+ *   in an uint32_t. This value will be the RX queue size of the
```
You're returning an int64_t.  Why are you telling the user they can store an unsigned integer in a container that COULD conceivably be out side of the range?

Also, I think it should read "in a uint32_t" versus "in an uint32_t"

bmagistro replied:
On success it can be safely stored in a uint32_t.  error codes are negative values.  i can't return an error code and support  the full range of the field using a int32_t so the size is bumped to a int64_t to allow negative error codes and the full range of the field on success.

ttosbt replied:
Understood, I'd still fix the English in there to use `a` versus `an` properly.

bmagistro replied:
feel like I'm missing something on this.  `an unsigned integer..` vs `a unsigned integer...`?

ttosbt replied:
In general "a" goes before words that begin with consonants. "An" goes before words that begin with vowels: an apricot. an egg.

https://owl.purdue.edu/owl/general_writing/grammar/articles_a_versus_an.html

ttosbt commented on lib/l4p/tle_udp.h:
```diff
@@ -98,6 +98,19 @@ int
+ *   On success, the value will be >= 0 and can safely be stored
+ *   in an uint32_t. This value will be the RX queue size of the
+ *   stream.
+ *   On error, the value will be < 0 and be one of the following:
```
You say "one of the following" yet only list one possible value...I'd revise this for clarity.

bmagistro replied:
...there are a number of functions that bubble up error codes that are not actually captured in the doxy like comments.  its one of my biggest gripes about this.  there are also a number of places where it expands to something like `-(-err_code)` but says errors will be negative...

ttosbt replied:
Haha, damn programmers.  :^)  I reckon that it is what it is on this one then.

stefan-baranoff commented on lib/l4p/udp_rxtx.c:
```diff
@@ -108,6 +108,33 @@ rx_stream(struct tle_udp_stream *s, void *mb[], struct rte_mbuf *rp[],
+
+	if (s->rx.ev != NULL)
+		tle_event_raise(s->rx.ev);
+	// unclear what purpose this serves since we arent acquiring a stream
```
I'd remove the comments here, I'm assuming this was copied from TCP where there is an acquire somewhere?

bmagistro replied:
this was based on `tle_udp_rx_bulk` https://github.com/Trinity-Cyber-Inc/tldk/blob/1dac0c24f33738f7f31c2ce4cb0bf98384f6954c/lib/l4p/udp_rxtx.c#L195

stefan-baranoff commented on lib/l4p/udp_stream.c:
```diff
@@ -276,6 +276,74 @@ tle_udp_stream_open(struct tle_ctx *ctx,
+		s->s.type = TLE_V6;
+		const struct in6_addr *pm;
+		pm = &tle_ipv6_none;
+		//fill_ipv6_am((const struct sockaddr_in6 *)&prm->local_addr,
```
Unnecessary comment?

bmagistro replied:
ipv6 has not been tested, this in theory is the correct implementation of that call...  i'd say only unnecessary if the implementation here was actually tested or fully reviewed and found to be accurate.  right now its 50/50 if it works and keeps the addresses in the right order

bmagistro replied:
v6 teste, comments removed

bmagistro committed:
l4p/tcp: multithreaded tle_tcp_process can lock up

We believe this is the correct set of step to follow this issue

1 - BE: RX a reset packet and start the rx_bulk call, we get past
rx_acquire_stream call but not all the way through the code path to
resetting the stream variables and critically not to the point of
setting type = TLE_VNUM. The rx_acquire_stream locks the stream, which
should prevent critical modifications.
2 - FE: Call tle_tcp_stream_abort, which calls stream_close_prolog with
operation ABORT_CLOSE
2a - FE: tle_tcp_stream_abort sets the stream operation for
stream_close_prolog to TLE_TCP_OP_CLOSE|TLE_TCP_OP_ABORT and then in
stream_close_prolog calls tcp_stream_down, pause processing in that
thread here.

This pause isn't just incidental, it is enforced by locking when this
race hits the right sequencing. The BE rx_acquire_stream forces a pause
in FE tcp_stream_down because it tries to lock as well, and is blocked
by the BE holding the lock.

3 - BE: continues processing the received RST through rx_stream which
will call into rx_rst; we're at a place the RST would pass check_seqn so
stream_term would get called
3a - BE: rx_rst calls stream_term which sets s->tcb.state =
TLE_TCP_ST_CLOSED; and because tle_tcp_stream_abort set
TLE_TCP_OP_CLOSE, the stream is enqueued to DR
3b - BE: back in rx_bulk the BE releases the lock by calling
tcp_stream_release allowing the FE to make progress
4 - FE: Picks back up in stream_close_prolog; now that the BE has set
stream state to closed, state <= TLE_TCP_ST_SYN_SENT is true in
stream_close_prolog so tcp_stream_reset(ctx, s); is called then return
0; causes tle_tcp_stream_abort to skip the tcp_stream_up
5 - BE: Back in the BE tle_tcp_process runs the death row list, which
our stream is now on and down, and calls tcp_stream_down then
tcp_stream_reset

----

The moved line (the stream down) puts the update to s->tcb.uop inside a
critical section protecting it from simultaneous update. When the FE
gets to step 2 it pauses and waits on the lock. The FE doesn't get to
2a. So the stream operation is NOT set to
TLE_TCP_OP_CLOSE|TLE_TCP_OP_ABORT yet which means in the BE in 3a the
check does NOT put the stream in the DR and instead raises an event.

Once the BE finishes up and unlocks the FE picks back up. THEN it
updates s->tcb.uop in stream_close_prolog and keeps processing and picks
up at 4. But step 5 is never run, because the sequencing did not trigger
death row enqueue.

Signed-off-by: Ben Magistro <koncept1@gmail.com>
Change-Id: Idffe9fca7ebf68f1fef38ecaa163e7c2f55e43f5
